File: lib/prepare.js
const SemanticReleaseError = require('@semantic-release/error');
const { readFile, writeFile } = require('node:fs/promises');
const { cargoExecutable } = require('./utils');

/**
 * @typedef {Object} PluginConfig
 * @property {string?} executable the executable that shall be called.
 * @property {Boolean?} check if check is executed.
 * @property {Boolean?} allFeatures if check is executed.
 * @property {Array<string>?} checkArgs additional arguments for the publish call.
 */

/**
 * Prepare the package for release (perform "cargo check" and set version number).
 *
 * @param {PluginConfig} pluginConfig
 * @param {import('semantic-release').Context} context
 */
module.exports = async ({ executable, allFeatures = false, check = true, checkArgs = [] }, { logger, nextRelease }) => {
  const { execa } = await import('execa');

  logger.info(`Write new release version (${nextRelease?.version}) into Cargo.toml.`);
  const tomlContent = await readFile('./Cargo.toml', 'utf8');
  await writeFile(
    './Cargo.toml',
    tomlContent.replace(/^version\s*=\s*(.*)/m, `version = "${nextRelease?.version}"`),
    'utf8'
  );

  if (check) {
    logger.info('Perform cargo check.');
    if (allFeatures && !checkArgs.includes('--all-features')) {
      checkArgs.push('--all-features');
    }

    const { stderr, exitCode } = await execa(cargoExecutable(executable), ['check', ...checkArgs]);
    if (exitCode !== 0) {
      throw new SemanticReleaseError('Cargo check failed', 'ECARGOCHECK', stderr);
    }
  }
};

File: src/Prepare.fs
module SemanticReleaseCargo.Prepare

open System.Text.RegularExpressions
open SemanticReleaseCargo.Config
open SemanticReleaseCargo.Errors
open SemanticReleaseCargo.ExternalApi
open SemanticReleaseCargo.SemanticRelease
open SemanticReleaseCargo.Utils
open SemanticReleaseCargo.Workspaces

let private versionRegex =
    Regex(@"^version\s*=\s*(.*)", RegexOptions.ECMAScript ||| RegexOptions.Multiline)

let prepare (api: IExternalApi) (config: PluginConfig) (context: PrepareContext) =
    async {
        let updateVersion crate = async {
            let path = composeManifestPath crate
            context.logger.info $"Write new release version ({context.nextRelease.version}) into {path}"
            let! tomlContent = api.readFile path

            let tomlContent =
                versionRegex.Replace(tomlContent, $"version = \"{context.nextRelease.version}\"")

            do! api.writeFile path tomlContent
        }

        runAsyncs <|
            match config.crates with
            | Some crates -> Array.map updateVersion crates
            | None ->  [| updateVersion "" |]

        if (true, config.check) ||> Option.defaultValue then
            context.logger.info "Run cargo check."

            let args = ([||], config.checkArgs) ||> Option.defaultValue
            let args = addPackageFlags config.crates context args "ECARGOCHECK"
            let allFeatures = (false, config.allFeatures) ||> Option.defaultValue

            let args =
                match (Array.contains "--all-features" args, allFeatures) with
                | false, true -> Array.append args [| "--all-features" |]
                | _ -> args

            let! _, err, exit = api.exec <| Array.append [| "check" |] args

            if exit <> 0 then
                context.logger.error $"Cargo check failed: {err}"
                raise (SemanticReleaseError("Cargo check failed.", "ECARGOCHECK", Some(err)))

        ()
    }
